### PR TITLE
Parser UNOC support + Encoder UNA segment support

### PR DESCRIPTION
**Pull request by homer8173**

### CODE REVIEW

File: src/EDI/Encoder.php [hunk 1]
```diff
@@ -1,25 +1,58 @@
...
 namespace EDI;
 
-class Encoder
-{
-    private $output='';
+class Encoder {
+
+    private $output = '';
+    private $UNAActive = false; // disable by default to preserve backward compatibility 
+    private $array = array();
```
Comment (sabas) [inline src/EDI/Encoder.php]: Could we use here a more specific name? perhaps $original_array

Comment (mtdavidson) [inline src/EDI/Encoder.php]: Also short array syntax would be good :)

File: src/EDI/Encoder.php [hunk 2]
```diff
@@ -31,41 +64,80 @@ public function encode($arr, $wrap = true)
...
+            $str .= $this->sep_data;
         }
-        $str .= "'";
+        $str .= $this->symb_end;
         return $str;
     }
 
-    private function escapeValue($str)
-    {
+//    private function escapeValue($str) {
```
Comment (sabas) [inline src/EDI/Encoder.php]: This commented out version is here to be tested?

File: src/EDI/Encoder.php [hunk 3]
```diff
@@ -1,25 +1,58 @@
...
+class Encoder {
+
+    private $output = '';
+    private $UNAActive = false; // disable by default to preserve backward compatibility 
+    private $array = array();
+
+    /**
+     * @var string : component separator character (default :)
+     */
+    private $sep_comp;
```
Comment (homer8173) [inline src/EDI/Encoder.php]: i made it symmetrical with Parser, that's were those $sep_xxx are comming from, but i agree that s not consistent.

File: src/EDI/Encoder.php [hunk 4]
```diff
@@ -31,41 +64,80 @@ public function encode($arr, $wrap = true)
...
-    {
-        return $this->output;
+    public function get() {
+        if ($this->UNAActive) {
+            return "UNA" . $this->sep_comp .
+                    $this->sep_data .
+                    $this->sep_dec .
+                    $this->symb_rel .
+                    $this->symb_rep .
+                    $this->symb_end . $this->output;
```
Comment (sabas) [inline src/EDI/Encoder.php]: Here we should consider the $wrap-ping

File: src/EDI/Encoder.php [hunk 5]
```diff
@@ -31,41 +64,80 @@ public function encode($arr, $wrap = true)
...
+            $this->sep_comp = $chars[0];
+            $this->sep_data = $chars[1];
+            $this->sep_dec = $chars[2];
+            $this->symb_rel = $chars[3];
+            $this->symb_rep = $chars[4];
+            $this->symb_end = $chars[5];
+            if ($this->output != '') {
+                $this->output = $this->encode($this->array);
+            }
+            return TRUE;
```
Comment (sabas) [inline src/EDI/Encoder.php]: TRUE / FALSE should be lowercased per PSR-2 standard

File: src/EDI/Encoder.php (hunk 1 above)
Comment (homer8173) [inline src/EDI/Encoder.php]: i'll do it

File: src/EDI/Encoder.php (hunk 2 above)
Comment (homer8173) [inline src/EDI/Encoder.php]: just a cancel ommited, i should rewrite this function, it's not clear with those escape chars

File: src/EDI/Parser.php [hunk 6]
```diff
@@ -40,16 +39,34 @@ class Parser
      * @var string : end character (default ')
      */
     private $symb_end;
+    /**
+     * @var string : encoding (default UNOA)
+     */
+    private $encoding;
     
+    private $encodintToStripChars = [
+        "UNOA" => "/[\x01-\x1F\x80-\xFF]/", // not as restrictive as it should be
```
Comment (sabas) [inline src/EDI/Parser.php]: If we set default UNOB, I think that perhaps the UNOA should be the restrictive version.

File: src/EDI/Parser.php [hunk 7]
```diff
@@ -40,16 +39,34 @@ class Parser
      * @var string : end character (default ')
      */
     private $symb_end;
+    /**
+     * @var string : encoding (default UNOA)
+     */
+    private $encoding;
     
+    private $encodintToStripChars = [
```
Comment (sabas) [inline src/EDI/Parser.php]: encodint? or encoding?

File: src/EDI/Parser.php (hunk 6 above)
Comment (homer8173) [inline src/EDI/Parser.php]: the risk is to filter UNOA as it should be and deprecate past code. that why i haven't updated UNOA  limitation

File: src/EDI/Parser.php (hunk 7 above)
Comment (homer8173) [inline src/EDI/Parser.php]: you are right, that should be encoding ...

File: src/EDI/Encoder.php (hunk 3 above)
Comment (sabas) [inline src/EDI/Encoder.php]: If they are all private variables we can do the change in this PR, otherwise I'd keep them separate changes.

**Comment by homer8173** [comment]
I let you check if it fit your requirements. 

Regards.

**Comment by sabas** [comment]
Merged, thanks @homer8173 !

**Pull request merged**